homeassistant/components/escea/climate.py

```diff
@@ -0,0 +1,222 @@
+
+from homeassistant.components.climate import ClimateEntity
+from homeassistant.components.climate.const import (
+    FAN_AUTO,
+    FAN_HIGH,
+    FAN_LOW,
+    HVAC_MODE_HEAT,
+    HVAC_MODE_OFF,
+    SUPPORT_FAN_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+)
+from homeassistant.const import ATTR_TEMPERATURE, PRECISION_WHOLE, TEMP_CELSIUS
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import ConfigType
+
+from .const import (
+    DATA_DISCOVERY_SERVICE,
+    DISPATCH_CONTROLLER_DISCONNECTED,
+    DISPATCH_CONTROLLER_DISCOVERED,
+    DISPATCH_CONTROLLER_RECONNECTED,
+    DISPATCH_CONTROLLER_UPDATE,
+    ESCEA,
+    ICON,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+_ESCEA_FAN_TO_HA = {
+    Controller.Fan.FLAME_EFFECT: FAN_LOW,
+    Controller.Fan.FAN_BOOST: FAN_HIGH,
+    Controller.Fan.AUTO: FAN_AUTO,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config: ConfigType, async_add_entities: Callable
+):
+    """Initialize an Escea Controller."""
+    disco = hass.data[DATA_DISCOVERY_SERVICE]
+
+    @callback
+    def init_controller(ctrl: Controller):
+        """Register the controller device."""
+
+        _LOGGER.info("Controller UID=%s discovered", ctrl.device_uid)
+
+        device = ControllerDevice(ctrl)
+        async_add_entities([device])
+
+    # create any components not yet created
+    for controller in disco.pi_disco.controllers.values():
+        init_controller(controller)
+
+    # connect to register any further components
+    async_dispatcher_connect(hass, DISPATCH_CONTROLLER_DISCOVERED, init_controller)
+
+    return True
+
+
+class ControllerDevice(ClimateEntity):
+    """Representation of Escea Controller."""
+
+    _attr_hvac_modes = [
+        HVAC_MODE_HEAT,
+        HVAC_MODE_OFF,
+    ]
+    _attr_supported_features = SUPPORT_TARGET_TEMPERATURE | SUPPORT_FAN_MODE
+    _attr_target_temperature_step = PRECISION_WHOLE
+    _attr_temperature_unit = TEMP_CELSIUS
+
+    _attr_precision = PRECISION_WHOLE
+    _attr_icon = ICON
+    _attr_should_poll = False
+
+    def __init__(self, controller: Controller) -> None:
+        """Initialise ControllerDevice."""
+        self._controller = controller
+
+        self._attr_min_temp = controller.min_temp
+        self._attr_max_temp = controller.max_temp
+
+        self._fan_to_pescea = {}
+        for fan in controller.Fan:
+            self._fan_to_pescea[_ESCEA_FAN_TO_HA[fan]] = fan
+        self._attr_fan_modes = list(self._fan_to_pescea)
+
+        self._attr_unique_id = controller.device_uid
+        self._attr_name = f"Escea Fireplace {self._attr_unique_id}"
+
+        self._attr_available = True
+
+    async def async_added_to_hass(self):
+        """Call on adding to hass.
+
+        Registers for connect/disconnect/update events
+        """
+
+        @callback
+        def controller_disconnected(ctrl: Controller, ex: Exception) -> None:
```

Comment:
Hi, thanks for the feedback, and I can see that it could help simplify further, but I think for the moment I'd rather not change the structure (which I'm confident in, as this integration has been working well for me for a couple of months now) until I am more confident with HA and python.

Reply:
Hi again, I've looked at this again, after a long break - I'm afraid I don't know how to make this improvement at the moment.